### PR TITLE
jit: support select instruction

### PR DESCRIPTION
**Pull request by mathetake**
part of #65

### CODE REVIEW

File: wasm/jit/jit_amd64.go [hunk 1]
```diff
@@ -1063,7 +1141,7 @@ func (b *amd64Builder) movConstToRegister(val int64, targetRegister int16) *obj.
 	return prog
 }
 
-func (b *amd64Builder) releaseRegister(loc *valueLocation) {
+func (b *amd64Builder) releaseRegisterToStack(loc *valueLocation) {
```
Comment (mathetake) [inline wasm/jit/jit_amd64.go:1144]: renamed to `releaseRegisterToStack` from `releaseRegister` to clarify!

**Comment by mathetake** [comment]
done writing test!

File: wasm/jit/jit_amd64.go [hunk 2]
```diff
@@ -807,27 +880,32 @@ func (b *amd64Builder) handleConstI64(o *wazeroir.OperationConstI64) error {
 	return nil
 }
 
-func (b *amd64Builder) moveStackToRegister(tp generalPurposeRegisterType, loc *valueLocation) error {
+func (b *amd64Builder) moveStackToRegisterWithAllocation(tp generalPurposeRegisterType, loc *valueLocation) error {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:883]: vs splitting into `b.allocateRegisterAndMarkUsed(tp, loc)` followed by `b.moveStackToRegister(loc)`? This function maybe is doing too many things to say at once..

File: wasm/jit/jit_amd64_test.go [hunk 3]
```diff
@@ -1481,3 +1481,134 @@ func TestAmd64Builder_handleUnreachable(t *testing.T) {
...
+		{x1OnRgister: false, x2OnRegister: true, selectX1: false, condValueOnCondRegister: true},
+		{x1OnRgister: false, x2OnRegister: false, selectX1: true, condValueOnCondRegister: true},
+		{x1OnRgister: false, x2OnRegister: false, selectX1: false, condValueOnCondRegister: true},
+	} {
+		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
+			builder := requireNewBuilder(t)
+			eng := newEngine()
+			builder.initializeReservedRegisters()
+			var x1, x2, c *valueLocation
+			if tc.x1OnRgister {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:1537]:
```suggestion
			if tc.x1OnRegister {
```

**Review by codefromthecrypt — APPROVED** [review]
Some style suggestions.

Also, not to address in this PR, but I think we'll need a less hairy way to setup tests at some point as the logic to test is a little hairy. It looks very similar to writing main code and less like assertions. It would be cool to have a test utility at some point to setup the preconditions and assertions in an easier to read way. granted that is easier said than done!

File: wasm/jit/jit_amd64.go [hunk 4]
```diff
@@ -564,6 +567,76 @@ func (b *amd64Builder) emitDropRange(r *wazeroir.InclusiveRange) error {
...
+		x2.register = tmpRegister
+		b.moveStackToRegister(x2)
+	}
+
+	//
+	// At this point x2's value is always on a register.
+	//
+
+	// Then release the value in the x2's register to the x1's stack position.
+	if x1.onRegister() {
```
[the remark this answers is not in the document]
Comment (mathetake) [reply wasm/jit/jit_amd64.go:618]: x1 is not consumed but modified with the new selected value. Will put more comments about that!

[the remark this answers is not in the document]
Comment (mathetake) [reply wasm/jit/jit_amd64.go]: https://github.com/tetratelabs/wazero/pull/71/commits/d2a2a328de50e7fe3e608fd32963b2680a9a60b8 done

File: wasm/jit/jit_amd64.go [hunk 5]
```diff
@@ -564,6 +567,76 @@ func (b *amd64Builder) emitDropRange(r *wazeroir.InclusiveRange) error {
 	return nil
 }
 
+func (b *amd64Builder) handleSelect() error {
```
[the remark this answers is not in the document]
Comment (mathetake) [reply wasm/jit/jit_amd64.go:576]: https://github.com/tetratelabs/wazero/pull/71/commits/0a1533aa20f8f5c9d63e0d16cd582259a7e2fd06 done!

**Comment by mathetake** [comment]
> Also, not to address in this PR, but I think we'll need a less hairy way to setup tests at some point as the logic to test is a little hairy. It looks very similar to writing main code and less like assertions. It would be cool to have a test utility at some point to setup the preconditions and assertions in an easier to read way. granted that is easier said than done!

Totally agree the test is getting hairy.. at some point before I finish all the impl of instructions, I would do some refactoring and add helpers to clean up!

File: wasm/jit/jit_amd64.go (hunk 2 above)
Comment (mathetake) [reply wasm/jit/jit_amd64.go:891]: haha yeah, I will leave as-is, but I will put some comments about this as TODO!

Comment (mathetake) [reply wasm/jit/jit_amd64.go:892]: https://github.com/tetratelabs/wazero/pull/71/commits/98fe7f43a3bcce6cbc91ae8887326c11bab710cb

**Comment by mathetake** [comment]
@codefromthecrypt thanks for the review! 👍

**Pull request merged**